controller_interface/include/controller_interface/async_function_handler.hpp

```diff
@@ -0,0 +1,263 @@
+   */
+  void join_async_update_thread()
+  {
+    if (is_running())
+    {
+      thread_.join();
+    }
+  }
+
+  /// Check if the async update thread is running
+  /**
+   * @return True if the async update thread is running, false otherwise
+   */
+  bool is_running() const { return thread_.joinable(); }
+
+  /// Check if the async update is triggered to stop the cycle
+  /**
+   * @return True if the async update is stopped, false otherwise
+   */
+  bool is_stopped() const
+  {
+    return (
+      async_update_stop_ ||
+      (is_initialized() &&
+       get_state_function_().id() != lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE));
+  }
+
+  /// Stops the async update thread
+  /**
+   * If the async method is running, it will notify the async thread to stop and then joins the
+   * async thread.
+   */
+  void stop_async_update()
+  {
+    if (is_running())
+    {
+      {
+        std::unique_lock<std::mutex> lock(async_mtx_);
+        async_update_stop_ = true;
+      }
+      async_update_condition_.notify_one();
+      thread_.join();
+    }
+  }
+
+  /// Initialize the async update thread
+  /**
+   * If the async update thread is not running, it will start the async update thread.
+   * If the async update thread is already configured and running, does nothing and return
+   * immediately.
+   */
+  void start_async_update_thread()
+  {
+    if (!is_initialized())
+    {
+      throw std::runtime_error("AsyncFunctionHandler: need to be initialized first!");
+    }
+    if (!thread_.joinable())
+    {
+      async_update_stop_ = false;
+      trigger_in_progress_ = false;
+      async_update_return_ = T();
+      thread_ = std::thread(
+        [this]() -> void
+        {
+          // \note There might be an concurrency issue with the get_state() call here. This mightn't
+          // be critical here as the state of the controller is not expected to change during the
+          // update cycle
+          while ((get_state_function_().id() == lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE ||
+                  get_state_function_().id() ==
+                    lifecycle_msgs::msg::State::TRANSITION_STATE_ACTIVATING) &&
+                 !async_update_stop_)
+          {
+            {
+              std::unique_lock<std::mutex> lock(async_mtx_);
+              async_update_condition_.wait(
+                lock, [this] { return trigger_in_progress_ || async_update_stop_; });
+              if (async_update_stop_)
+              {
+                trigger_in_progress_ = false;
+                cycle_end_condition_.notify_one();
+                break;
+              }
+              async_update_return_ = async_function_(current_update_time_, current_update_period_);
+              trigger_in_progress_ = false;
+            }
+            cycle_end_condition_.notify_one();
+          }
+          trigger_in_progress_ = false;
+          cycle_end_condition_.notify_one();
```

Comment:
Are these notifications needed?

Reply:
Probably, it is not needed. I'll remove it